File: axios_compromise_scanner.py
#!/usr/bin/env python3
"""
Standalone scanner for the March 2026 axios npm supply-chain compromise.

References used to build IOCs and checks:
- StepSecurity incident write-up (malicious versions + attack behavior)
- Huntress incident write-up (package hashes + host/network/file IOCs)

This script is intentionally standalone (stdlib only) and can be run as:

  python3 axios_compromise_scanner.py --root .

Exit codes:
  0 = no findings
  1 = findings detected
  2 = script/runtime error
"""

from __future__ import annotations

import argparse
import json
import os
import re
import sys
import tempfile
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Iterable, List

MALICIOUS_AXIOS_VERSIONS = {"1.14.1", "0.30.4"}
MALICIOUS_DEPENDENCY = "plain-crypto-js"
SUSPICIOUS_PLAIN_CRYPTO_VERSIONS = {"4.2.1"}

# Published IOC hashes from Huntress for malicious npm packages.
MALICIOUS_SHA1 = {
    "axios@1.14.1": "2553649f2322049666871cea80a5d0d6adc700ca",
    "axios@0.30.4": "d6f3f62fd3b9f5432f5782b62d8cfd5247d5ee71",
    "plain-crypto-js@4.2.1": "07d889e2dadce6f3910dcbc253317d28ca61c766",
}

NETWORK_IOCS = {
    "domains": ["sfrclak.com"],
    "ip_addresses": ["142.11.206.73"],
    "urls": ["http://sfrclak.com:8000/6202033"],
    "post_bodies": [
        "packages.npm.org/product0",
        "packages.npm.org/product1",
        "packages.npm.org/product2",
    ],
    "user_agent": "mozilla/4.0 (compatible; msie 8.0; windows nt 5.1; trident/4.0)",
}

FILE_IOCS = [
    "/Library/Caches/com.apple.act.mond",
    r"%PROGRAMDATA%\\wt.exe",
    r"%PROGRAMDATA%\\system.bat",
    r"%TEMP%\\6202033.vbs",
    r"%TEMP%\\6202033.ps1",
    "/tmp/ld.py",
]

LOCKFILE_CANDIDATES = {
    "package-lock.json",
    "npm-shrinkwrap.json",
    "yarn.lock",
    "pnpm-lock.yaml",
}

SKIP_DIRS = {
    ".git",
    ".hg",
    ".svn",
    ".next",
    ".turbo",
    ".cache",
    "dist",
    "build",
    "coverage",
}


@dataclass
class Finding:
    severity: str
    category: str
    path: str
    summary: str
    details: str


@dataclass
class ScanReport:
    scanned_root: str
    findings: List[Finding]

    @property
    def compromised(self) -> bool:
        return any(f.severity in {"critical", "high"} for f in self.findings)

    def to_json(self) -> str:
        return json.dumps(
            {
                "scanned_root": self.scanned_root,
                "compromised": self.compromised,
                "finding_count": len(self.findings),
                "findings": [asdict(f) for f in self.findings],
                "iocs": {
                    "malicious_versions": {
                        "axios": sorted(MALICIOUS_AXIOS_VERSIONS),
                        MALICIOUS_DEPENDENCY: sorted(SUSPICIOUS_PLAIN_CRYPTO_VERSIONS),
                    },
                    "malicious_sha1": MALICIOUS_SHA1,
                    "network": NETWORK_IOCS,
                    "filesystem": FILE_IOCS,
                },
            },
            indent=2,
        )


def iter_files(root: Path) -> Iterable[Path]:
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in SKIP_DIRS]
        for name in filenames:
            yield Path(dirpath) / name


def safe_read(path: Path) -> str:
    try:
        return path.read_text(encoding="utf-8", errors="replace")
    except Exception:
        return ""


def check_lockfile_for_versions(path: Path, findings: List[Finding]) -> None:
    text = safe_read(path)
    if not text:
        return

    def add_axios_finding(version: str) -> None:
        findings.append(
            Finding(
                severity="critical",
                category="dependency",
                path=str(path),
                summary=f"Potential malicious axios version referenced: {version}",
                details=(
                    "Lockfile contains a known malicious axios version from the March 2026 compromise. "
                    "Treat this environment as potentially compromised and rotate credentials."
                ),
            )
        )

    # Prefer structured parsing for JSON lockfiles where available.
    if path.suffix == ".json":
        try:
            parsed = json.loads(text)
        except json.JSONDecodeError:
            parsed = None

        if isinstance(parsed, dict):
            def recurse(obj: object) -> None:
                if isinstance(obj, dict):
                    if obj.get("name") == "axios" and str(obj.get("version", "")) in MALICIOUS_AXIOS_VERSIONS:
                        add_axios_finding(str(obj["version"]))
                    for key, value in obj.items():
                        if key == "axios" and isinstance(value, dict):
                            version = str(value.get("version", "")).strip()
                            if version in MALICIOUS_AXIOS_VERSIONS:
                                add_axios_finding(version)
                        recurse(value)
                elif isinstance(obj, list):
                    for item in obj:
                        recurse(item)

            recurse(parsed)

    for version in MALICIOUS_AXIOS_VERSIONS:
        pattern = rf"axios(?:@|\s|/)?{re.escape(version)}"
        if re.search(pattern, text):
            add_axios_finding(version)

    if re.search(r"plain-crypto-js", text):
        sev = "critical" if re.search(r"plain-crypto-js[^\n]*4\.2\.1", text) else "high"
        findings.append(
            Finding(
                severity=sev,
                category="dependency",
                path=str(path),
                summary="Suspicious plain-crypto-js dependency reference detected",
                details=(
                    "plain-crypto-js appeared in the axios compromise chain and should not be present in normal axios installs."
                ),
            )
            )


def check_generic_json_for_indicators(path: Path, findings: List[Finding]) -> None:
    text = safe_read(path)
    if not text:
        return

    matched = False

    # Structured parse for JSON content.
    try:
        parsed = json.loads(text)
    except json.JSONDecodeError:
        parsed = None

    if isinstance(parsed, (dict, list)):
        def recurse(obj: object) -> None:
            nonlocal matched
            if isinstance(obj, dict):
                if obj.get("name") == "axios" and str(obj.get("version", "")) in MALICIOUS_AXIOS_VERSIONS:
                    version = str(obj.get("version", "")).strip()
                    findings.append(
                        Finding(
                            severity="critical",
                            category="json-indicator",
                            path=str(path),
                            summary=f"JSON indicator references malicious axios version: {version}",
                            details="Structured JSON parsing identified a known malicious axios version field.",
                        )
                    )
                    matched = True
                for key, value in obj.items():
                    if key == "axios" and isinstance(value, dict):
                        version = str(value.get("version", "")).strip()
                        if version in MALICIOUS_AXIOS_VERSIONS:
                            findings.append(
                                Finding(
                                    severity="critical",
                                    category="json-indicator",
                                    path=str(path),
                                    summary=f"JSON indicator references malicious axios version: {version}",
                                    details="Structured JSON parsing identified axios object with malicious version.",
                                )
                            )
                            matched = True
                    recurse(value)
            elif isinstance(obj, list):
                for item in obj:
                    recurse(item)

        recurse(parsed)

    # Regex pass is always applied as an additional safety net.
    regex_hit = False
    for version in MALICIOUS_AXIOS_VERSIONS:
        if re.search(rf"axios(?:@|\s|/)?{re.escape(version)}", text) or re.search(
            rf"\"axios\"\\s*:\\s*\\{{[^\\}}]*\"version\"\\s*:\\s*\"{re.escape(version)}\"",
            text,
        ):
            regex_hit = True
            findings.append(
                Finding(
                    severity="critical",
                    category="json-indicator",
                    path=str(path),
                    summary=f"Regex indicator references malicious axios version: {version}",
                    details="Regex scanning of JSON text matched a known malicious axios version pattern.",
                )
            )

    if re.search(r"plain-crypto-js", text):
        regex_hit = True
        findings.append(
            Finding(
                severity="high",
                category="json-indicator",
                path=str(path),
                summary="Regex indicator references plain-crypto-js in JSON file",
                details="Regex scanning found plain-crypto-js token in JSON content; investigate dependency chain.",
            )
        )

    # Avoid silent no-op for JSON files that had no indicators.
    _ = matched or regex_hit


def check_installed_node_modules(root: Path, findings: List[Finding]) -> None:
    axios_pkg = root / "node_modules" / "axios" / "package.json"
    if axios_pkg.exists():
        try:
            data = json.loads(safe_read(axios_pkg) or "{}")
        except json.JSONDecodeError:
            data = {}
        version = str(data.get("version", "")).strip()
        if version in MALICIOUS_AXIOS_VERSIONS:
            findings.append(
                Finding(
                    severity="critical",
                    category="installed-package",
                    path=str(axios_pkg),
                    summary=f"Installed axios version is malicious: {version}",
                    details="Known compromised version discovered directly in node_modules.",
                )
            )

    plain_crypto_dir = root / "node_modules" / MALICIOUS_DEPENDENCY
    if plain_crypto_dir.exists():
        details = (
            "Directory exists. Huntress notes presence of node_modules/plain-crypto-js can indicate compromise, "
            "even if package contents look benign after self-cleanup."
        )
        pkg = plain_crypto_dir / "package.json"
        version = ""
        if pkg.exists():
            try:
                version = str(json.loads(safe_read(pkg) or "{}").get("version", "")).strip()
            except json.JSONDecodeError:
                pass
        if version in SUSPICIOUS_PLAIN_CRYPTO_VERSIONS:
            sev = "critical"
            details += f" package.json reports suspicious version {version}."
        else:
            sev = "high"
            if version:
                details += f" package.json reports version {version}; this can be spoofed post-infection."
        findings.append(
            Finding(
                severity=sev,
                category="installed-package",
                path=str(plain_crypto_dir),
                summary="Suspicious dependency directory detected: node_modules/plain-crypto-js",
                details=details,
            )
        )


def check_host_iocs(findings: List[Finding]) -> None:
    expanded_paths = [
        Path("/Library/Caches/com.apple.act.mond"),
        Path("/tmp/ld.py"),
    ]

    if os.name == "nt":
        for env_var, leaf in [
            ("PROGRAMDATA", "wt.exe"),
            ("PROGRAMDATA", "system.bat"),
            ("TEMP", "6202033.vbs"),
            ("TEMP", "6202033.ps1"),
        ]:
            base = os.environ.get(env_var)
            if base:
                expanded_paths.append(Path(base) / leaf)

    for p in expanded_paths:
        if p.exists():
            findings.append(
                Finding(
                    severity="critical",
                    category="host-artifact",
                    path=str(p),
                    summary="Known axios-compromise filesystem IOC detected",
                    details="IOC path exists on disk and maps to published malware artifacts.",
                )
            )


def run_scan(root: Path) -> ScanReport:
    findings: List[Finding] = []

    for path in iter_files(root):
        if path.name in LOCKFILE_CANDIDATES:
            check_lockfile_for_versions(path, findings)
        elif path.suffix == ".json":
            check_generic_json_for_indicators(path, findings)

    check_installed_node_modules(root, findings)
    check_host_iocs(findings)

    return ScanReport(scanned_root=str(root.resolve()), findings=findings)


def print_human_report(report: ScanReport) -> None:
    print("=" * 78)
    print("Axios 2026 Supply-Chain Compromise Scanner")
    print("=" * 78)
    print(f"Scanned root: {report.scanned_root}")
    print(f"Findings: {len(report.findings)}")
    print(f"Compromised (heuristic): {'YES' if report.compromised else 'NO'}")

    if report.findings:
        print("\nFindings:")
        for idx, finding in enumerate(report.findings, start=1):
            print(f"  {idx}. [{finding.severity.upper()}] {finding.summary}")
            print(f"     Category: {finding.category}")
            print(f"     Path:     {finding.path}")
            print(f"     Details:  {finding.details}")

    print("\nRecommended response if findings are present:")
    print("  1) Assume credential exposure on affected hosts; rotate secrets immediately.")
    print("  2) Rebuild affected systems from known-good images (avoid in-place cleanup).")
    print("  3) Pin axios to safe versions (1.14.0 or 0.30.3) and reinstall with --ignore-scripts.")
    print("  4) Block network IOC: sfrclak.com / 142.11.206.73 / port 8000.")


def parse_args(argv: List[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Standalone scanner for the March 2026 axios npm compromise."
    )
    parser.add_argument(
        "--root",
        default=".",
        help="Root directory to scan for lockfiles/node_modules (default: current dir)",
    )
    parser.add_argument(
        "--json",
        action="store_true",
        help="Print machine-readable JSON report",
    )
    parser.add_argument(
        "--self-test",
        action="store_true",
        help="Run built-in dummy payload tests and exit",
    )
    return parser.parse_args(argv)


def run_self_test() -> int:
    test_results = []

    def record(name: str, passed: bool, details: str = "") -> None:
        test_results.append((name, passed, details))

    with tempfile.TemporaryDirectory(prefix="axios-scan-test-") as tmp:
        root = Path(tmp)

        # Test 1: clean workspace should not be marked compromised.
        clean = run_scan(root)
        record(
            "clean_workspace",
            clean.compromised is False and len(clean.findings) == 0,
            f"findings={len(clean.findings)} compromised={clean.compromised}",
        )

        # Test 2: malicious axios version in lockfile should be detected.
        lock = root / "package-lock.json"
        lock.write_text(
            json.dumps({"dependencies": {"axios": {"version": "1.14.1"}}}),
            encoding="utf-8",
        )
        compromised = run_scan(root)
        has_axios = any("axios version" in f.summary.lower() for f in compromised.findings)
        record(
            "lockfile_malicious_axios",
            compromised.compromised is True and has_axios,
            f"findings={len(compromised.findings)}",
        )

        # Test 3: suspicious dependency directory should be detected.
        plain_crypto_pkg = root / "node_modules" / "plain-crypto-js"
        plain_crypto_pkg.mkdir(parents=True, exist_ok=True)
        (plain_crypto_pkg / "package.json").write_text(
            json.dumps({"name": "plain-crypto-js", "version": "4.2.1"}),
            encoding="utf-8",
        )
        compromised2 = run_scan(root)
        has_plain_crypto = any("plain-crypto-js" in f.summary for f in compromised2.findings)
        record(
            "node_modules_plain_crypto",
            compromised2.compromised is True and has_plain_crypto,
            f"findings={len(compromised2.findings)}",
        )

    print("Self-test results:")
    failures = 0
    for name, passed, details in test_results:
        status = "PASS" if passed else "FAIL"
        print(f"  - {name}: {status} {details}")
        if not passed:
            failures += 1
    print(f"Summary: {len(test_results) - failures}/{len(test_results)} tests passed")
    return 0 if failures == 0 else 2


def main(argv: List[str]) -> int:
    try:
        args = parse_args(argv)
        if args.self_test:
            return run_self_test()

        root = Path(args.root)
        if not root.exists() or not root.is_dir():
            print(f"[error] --root does not exist or is not a directory: {root}", file=sys.stderr)
            return 2

        report = run_scan(root)

        if args.json:
            print(report.to_json())
        else:
            print_human_report(report)

        return 1 if report.compromised else 0
    except KeyboardInterrupt:
        print("\n[error] interrupted", file=sys.stderr)
        return 2
    except Exception as exc:
        print(f"[error] scanner failed: {exc}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: axios_compromise_scanner.js
#!/usr/bin/env node
'use strict';

/**
 * Standalone scanner for the March 2026 axios npm supply-chain compromise.
 *
 * Usage:
 *   node axios_compromise_scanner.js --root .
 *   node axios_compromise_scanner.js --root . --json
 *   node axios_compromise_scanner.js --self-test
 *
 * Exit codes:
 *   0 = no findings / successful self-test
 *   1 = findings detected
 *   2 = runtime error / failed self-test
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const MALICIOUS_AXIOS_VERSIONS = new Set(['1.14.1', '0.30.4']);
const MALICIOUS_DEPENDENCY = 'plain-crypto-js';
const SUSPICIOUS_PLAIN_CRYPTO_VERSIONS = new Set(['4.2.1']);

const MALICIOUS_SHA1 = {
  'axios@1.14.1': '2553649f2322049666871cea80a5d0d6adc700ca',
  'axios@0.30.4': 'd6f3f62fd3b9f5432f5782b62d8cfd5247d5ee71',
  'plain-crypto-js@4.2.1': '07d889e2dadce6f3910dcbc253317d28ca61c766',
};

const NETWORK_IOCS = {
  domains: ['sfrclak.com'],
  ip_addresses: ['142.11.206.73'],
  urls: ['http://sfrclak.com:8000/6202033'],
  post_bodies: ['packages.npm.org/product0', 'packages.npm.org/product1', 'packages.npm.org/product2'],
  user_agent: 'mozilla/4.0 (compatible; msie 8.0; windows nt 5.1; trident/4.0)',
};

const FILE_IOCS = [
  '/Library/Caches/com.apple.act.mond',
  '%PROGRAMDATA%\\wt.exe',
  '%PROGRAMDATA%\\system.bat',
  '%TEMP%\\6202033.vbs',
  '%TEMP%\\6202033.ps1',
  '/tmp/ld.py',
];

const LOCKFILE_CANDIDATES = new Set(['package-lock.json', 'npm-shrinkwrap.json', 'yarn.lock', 'pnpm-lock.yaml']);
const SKIP_DIRS = new Set(['.git', '.hg', '.svn', '.next', '.turbo', '.cache', 'dist', 'build', 'coverage']);

function finding(severity, category, fpath, summary, details) {
  return { severity, category, path: fpath, summary, details };
}

function safeRead(filePath) {
  try {
    return fs.readFileSync(filePath, 'utf8');
  } catch (_err) {
    return '';
  }
}

function* iterFiles(root) {
  const stack = [root];
  while (stack.length) {
    const current = stack.pop();
    let entries;
    try {
      entries = fs.readdirSync(current, { withFileTypes: true });
    } catch (_err) {
      continue;
    }

    for (const entry of entries) {
      const full = path.join(current, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIRS.has(entry.name)) stack.push(full);
      } else if (entry.isFile()) {
        yield full;
      }
    }
  }
}

function addAxiosFinding(findings, fpath, version, category = 'dependency') {
  findings.push(
    finding(
      'critical',
      category,
      fpath,
      `Potential malicious axios version referenced: ${version}`,
      'Known malicious axios version from the March 2026 compromise detected. Treat host as potentially compromised and rotate credentials.'
    )
  );
}

function checkLockfileForVersions(filePath, findings) {
  const text = safeRead(filePath);
  if (!text) return;

  if (filePath.endsWith('.json')) {
    let parsed = null;
    try {
      parsed = JSON.parse(text);
    } catch (_err) {
      parsed = null;
    }

    const walk = (obj) => {
      if (!obj || typeof obj !== 'object') return;
      if (Array.isArray(obj)) {
        for (const item of obj) walk(item);
        return;
      }

      if (obj.name === 'axios' && MALICIOUS_AXIOS_VERSIONS.has(String(obj.version || '').trim())) {
        addAxiosFinding(findings, filePath, String(obj.version).trim());
      }

      for (const [k, v] of Object.entries(obj)) {
        if (k === 'axios' && v && typeof v === 'object' && MALICIOUS_AXIOS_VERSIONS.has(String(v.version || '').trim())) {
          addAxiosFinding(findings, filePath, String(v.version).trim());
        }
        walk(v);
      }
    };

    if (parsed && typeof parsed === 'object') walk(parsed);
  }

  for (const version of MALICIOUS_AXIOS_VERSIONS) {
    const rx = new RegExp(`axios(?:@|\\s|/)?${version.replace('.', '\\.').replace('.', '\\.')}`);
    if (rx.test(text)) addAxiosFinding(findings, filePath, version);
  }

  if (/plain-crypto-js/.test(text)) {
    const sev = /plain-crypto-js[^\n]*4\.2\.1/.test(text) ? 'critical' : 'high';
    findings.push(
      finding(
        sev,
        'dependency',
        filePath,
        'Suspicious plain-crypto-js dependency reference detected',
        'plain-crypto-js appeared in the axios compromise chain and should not be present in normal axios installs.'
      )
    );
  }
}

function checkGenericJsonForIndicators(filePath, findings) {
  const text = safeRead(filePath);
  if (!text) return;

  let parsed = null;
  try {
    parsed = JSON.parse(text);
  } catch (_err) {
    parsed = null;
  }

  const walk = (obj) => {
    if (!obj || typeof obj !== 'object') return;
    if (Array.isArray(obj)) {
      for (const item of obj) walk(item);
      return;
    }

    if (obj.name === 'axios' && MALICIOUS_AXIOS_VERSIONS.has(String(obj.version || '').trim())) {
      findings.push(
        finding(
          'critical',
          'json-indicator',
          filePath,
          `JSON indicator references malicious axios version: ${String(obj.version).trim()}`,
          'Structured JSON parsing identified a malicious axios version field.'
        )
      );
    }

    for (const [k, v] of Object.entries(obj)) {
      if (k === 'axios' && v && typeof v === 'object' && MALICIOUS_AXIOS_VERSIONS.has(String(v.version || '').trim())) {
        findings.push(
          finding(
            'critical',
            'json-indicator',
            filePath,
            `JSON indicator references malicious axios version: ${String(v.version).trim()}`,
            'Structured JSON parsing identified axios object with malicious version.'
          )
        );
      }
      walk(v);
    }
  };

  if (parsed && typeof parsed === 'object') walk(parsed);

  for (const version of MALICIOUS_AXIOS_VERSIONS) {
    const rx1 = new RegExp(`axios(?:@|\\s|/)?${version.replace('.', '\\.').replace('.', '\\.')}`);
    const rx2 = new RegExp(`"axios"\\s*:\\s*\\{[^\\}]*"version"\\s*:\\s*"${version.replace('.', '\\.').replace('.', '\\.')}"`);
    if (rx1.test(text) || rx2.test(text)) {
      findings.push(
        finding(
          'critical',
          'json-indicator',
          filePath,
          `Regex indicator references malicious axios version: ${version}`,
          'Regex scanning of JSON text matched a malicious axios version pattern.'
        )
      );
    }
  }

  if (/plain-crypto-js/.test(text)) {
    findings.push(
      finding(
        'high',
        'json-indicator',
        filePath,
        'Regex indicator references plain-crypto-js in JSON file',
        'Regex scanning found plain-crypto-js token in JSON content; investigate dependency chain.'
      )
    );
  }
}

function checkInstalledNodeModules(root, findings) {
  // Recursively inspect all node_modules directories under root so scans work
  // for full home-directory scans and monorepos.
  for (const filePath of iterFiles(root)) {
    if (!filePath.endsWith(path.join('node_modules', 'axios', 'package.json'))) continue;
    let version = '';
    try {
      version = String(JSON.parse(safeRead(filePath) || '{}').version || '').trim();
    } catch (_err) {
      version = '';
    }
    if (MALICIOUS_AXIOS_VERSIONS.has(version)) {
      findings.push(
        finding('critical', 'installed-package', filePath, `Installed axios version is malicious: ${version}`, 'Known compromised version discovered directly in nested node_modules.')
      );
    }
  }

  for (const filePath of iterFiles(root)) {
    if (!filePath.endsWith(path.join('node_modules', MALICIOUS_DEPENDENCY, 'package.json'))) continue;
    const plainCryptoDir = path.dirname(filePath);
    let details = 'Directory exists. Huntress notes presence of node_modules/plain-crypto-js can indicate compromise, even if package contents look benign after self-cleanup.';
    let sev = 'high';
    try {
      const version = String(JSON.parse(safeRead(filePath) || '{}').version || '').trim();
      if (SUSPICIOUS_PLAIN_CRYPTO_VERSIONS.has(version)) {
        sev = 'critical';
        details += ` package.json reports suspicious version ${version}.`;
      } else if (version) {
        details += ` package.json reports version ${version}; this can be spoofed post-infection.`;
      }
    } catch (_err) {
      // noop
    }

    findings.push(
      finding(sev, 'installed-package', plainCryptoDir, 'Suspicious dependency directory detected: node_modules/plain-crypto-js', details)
    );
  }
}

function checkHostIocs(findings) {
  const expanded = ['/Library/Caches/com.apple.act.mond', '/tmp/ld.py'];

  if (process.platform === 'win32') {
    if (process.env.PROGRAMDATA) {
      expanded.push(path.join(process.env.PROGRAMDATA, 'wt.exe'));
      expanded.push(path.join(process.env.PROGRAMDATA, 'system.bat'));
    }
    if (process.env.TEMP) {
      expanded.push(path.join(process.env.TEMP, '6202033.vbs'));
      expanded.push(path.join(process.env.TEMP, '6202033.ps1'));
    }
  }

  for (const p of expanded) {
    if (fs.existsSync(p)) {
      findings.push(
        finding('critical', 'host-artifact', p, 'Known axios-compromise filesystem IOC detected', 'IOC path exists on disk and maps to published malware artifacts.')
      );
    }
  }
}

function runScan(root) {
  const findings = [];
  for (const filePath of iterFiles(root)) {
    const base = path.basename(filePath);
    if (LOCKFILE_CANDIDATES.has(base)) {
      checkLockfileForVersions(filePath, findings);
    } else if (filePath.endsWith('.json')) {
      checkGenericJsonForIndicators(filePath, findings);
    }
  }

  checkInstalledNodeModules(root, findings);
  checkHostIocs(findings);

  return {
    scanned_root: path.resolve(root),
    findings,
    compromised: findings.some((f) => f.severity === 'critical' || f.severity === 'high'),
    iocs: {
      malicious_versions: {
        axios: Array.from(MALICIOUS_AXIOS_VERSIONS).sort(),
        [MALICIOUS_DEPENDENCY]: Array.from(SUSPICIOUS_PLAIN_CRYPTO_VERSIONS).sort(),
      },
      malicious_sha1: MALICIOUS_SHA1,
      network: NETWORK_IOCS,
      filesystem: FILE_IOCS,
    },
  };
}

function printHumanReport(report) {
  console.log('==============================================================================');
  console.log('Axios 2026 Supply-Chain Compromise Scanner (Node.js)');
  console.log('==============================================================================');
  console.log(`Scanned root: ${report.scanned_root}`);
  console.log(`Findings: ${report.findings.length}`);
  console.log(`Compromised (heuristic): ${report.compromised ? 'YES' : 'NO'}`);

  if (report.findings.length) {
    console.log('\nFindings:');
    report.findings.forEach((f, idx) => {
      console.log(`  ${idx + 1}. [${String(f.severity || '').toUpperCase()}] ${f.summary}`);
      console.log(`     Category: ${f.category}`);
      console.log(`     Path:     ${f.path}`);
      console.log(`     Details:  ${f.details}`);
    });
  }

  console.log('\nRecommended response if findings are present:');
  console.log('  1) Assume credential exposure on affected hosts; rotate secrets immediately.');
  console.log('  2) Rebuild affected systems from known-good images (avoid in-place cleanup).');
  console.log('  3) Pin axios to safe versions (1.14.0 or 0.30.3) and reinstall with --ignore-scripts.');
  console.log('  4) Block network IOC: sfrclak.com / 142.11.206.73 / port 8000.');
}

function parseArgs(argv) {
  const args = { root: '.', json: false, selfTest: false };
  for (let i = 0; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === '--json') args.json = true;
    else if (a === '--self-test') args.selfTest = true;
    else if (a === '--root') {
      const v = argv[i + 1];
      if (!v) throw new Error('--root requires a path');
      args.root = v;
      i += 1;
    } else {
      throw new Error(`Unknown argument: ${a}`);
    }
  }
  return args;
}

function runSelfTest() {
  const results = [];
  const record = (name, passed, details = '') => results.push({ name, passed, details });

  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'axios-scan-test-node-'));
  try {
    const clean = runScan(tmp);
    record('clean_workspace', clean.compromised === false && clean.findings.length === 0, `findings=${clean.findings.length}`);

    const lock = path.join(tmp, 'package-lock.json');
    fs.writeFileSync(lock, JSON.stringify({ dependencies: { axios: { version: '1.14.1' } } }), 'utf8');
    const compromised = runScan(tmp);
    record('lockfile_malicious_axios', compromised.compromised === true && compromised.findings.some((f) => /axios version/i.test(f.summary)), `findings=${compromised.findings.length}`);

    const plainDir = path.join(tmp, 'node_modules', 'plain-crypto-js');
    fs.mkdirSync(plainDir, { recursive: true });
    fs.writeFileSync(path.join(plainDir, 'package.json'), JSON.stringify({ name: 'plain-crypto-js', version: '4.2.1' }), 'utf8');
    const compromised2 = runScan(tmp);
    record('node_modules_plain_crypto', compromised2.compromised === true && compromised2.findings.some((f) => f.summary.includes('plain-crypto-js')), `findings=${compromised2.findings.length}`);

    const nestedAxiosDir = path.join(tmp, 'users', 'example', 'project', 'node_modules', 'axios');
    fs.mkdirSync(nestedAxiosDir, { recursive: true });
    fs.writeFileSync(path.join(nestedAxiosDir, 'package.json'), JSON.stringify({ name: 'axios', version: '0.30.4' }), 'utf8');
    const compromised3 = runScan(tmp);
    record(
      'nested_node_modules_recursive_scan',
      compromised3.compromised === true && compromised3.findings.some((f) => /installed axios version is malicious/i.test(f.summary)),
      `findings=${compromised3.findings.length}`
    );

    console.log('Self-test results:');
    let failures = 0;
    for (const r of results) {
      console.log(`  - ${r.name}: ${r.passed ? 'PASS' : 'FAIL'} ${r.details}`);
      if (!r.passed) failures += 1;
    }
    console.log(`Summary: ${results.length - failures}/${results.length} tests passed`);

    return failures === 0 ? 0 : 2;
  } finally {
    fs.rmSync(tmp, { recursive: true, force: true });
  }
}

function main(argv) {
  try {
    const args = parseArgs(argv);
    if (args.selfTest) return runSelfTest();

    const root = args.root;
    if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
      console.error(`[error] --root does not exist or is not a directory: ${root}`);
      return 2;
    }

    const report = runScan(root);
    if (args.json) {
      console.log(JSON.stringify({ ...report, finding_count: report.findings.length }, null, 2));
    } else {
      printHumanReport(report);
    }

    return report.compromised ? 1 : 0;
  } catch (err) {
    console.error(`[error] scanner failed: ${err.message}`);
    return 2;
  }
}

process.exitCode = main(process.argv.slice(2));
